l3kernel/l3keys.dtx

```diff
@@ -1519,21 +1519,21 @@
 % \subsection{The key defining mechanism}
 %
 % \begin{macro}{\keys_define:nn}
-% \begin{macro}{\@@_define:nnn, \@@_define:onn}
+% \begin{macro}{\@@_define:nnn}
 %   The public function for definitions is just a wrapper for the lower
 %   level mechanism, more or less. The outer function is designed to
 %   keep a track of the current module, to allow safe nesting. The module is set
-%   removing any leading |/| (which is not needed here).
+%   removing any leading |/| (which is not needed here). The first argument to
+%   |\@@_define:nnn| should always be a string.
 %    \begin{macrocode}
 \cs_new_protected:Npn \keys_define:nn
-  { \@@_define:onn \l_@@_module_str }
+  { \exp_args:No \@@_define:nnn \l_@@_module_str }
 \cs_new_protected:Npn \@@_define:nnn #1#2#3
   {
-    \str_set:Nx \l_@@_module_str { \@@_trim_spaces:n {#2} }
+    \cs_set_nopar:Npx \l_@@_module_str { \@@_trim_spaces:n {#2} }
     \keyval_parse:NNn \@@_define:n \@@_define:nn {#3}
-    \str_set:Nn \l_@@_module_str {#1}
+    \cs_set_nopar:Npn \l_@@_module_str {#1}
   }
-\cs_generate_variant:Nn \@@_define:nnn { o }
```

Comment:
I'm ok with expanding `\@@_define:onn` because it is only used there and nowhere else.  This is not so much a speed optimization as an optimization of not needlessly introducing csnames.

The `\cs_set_nopar:Npx` speed optimizations are not worth the uglier code: this is typically done once per `\keys_define:nn`, hence a few times per package.  Optimizations concerning every key (like your earlier optimizations of `\keyval_parse:NNn` are more important.

On the other hand, maybe we should have a more general way to "set a tl variable to some value then restore the old value" as it is done here and in many other places in l3keys.
```
\cs_new_protected:Npn \keys_define:nn #1#2
  {
    \__kernel_with_tl_set:Nxn \l_@@_module_str
      { \@@_trim_spaces:n {#2} }
      { \keyval_parse:NNn \@@_define:n \@@_define:nn {#3} }
  }
\cs_new_protected:Npn \__kernel_with_tl_set:Nxn #1
  { \exp_args:No \__kernel_with_tl_set:nNxn {#1} #1 }
\cs_new_protected:Npn \__kernel_with_tl_set:nNxn #1#2#3#4
  {
    \tl_set:Nx #2 {#3}
    #4
    \tl_set:Nn #2 {#1}
  }
```

Reply:
I like the idea of `\__kernel_with_tl_set:Nxn` to unify this code, but that should go to another issue/PR.